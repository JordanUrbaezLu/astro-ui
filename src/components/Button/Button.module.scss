@use "../../utility/globals";

.button {
  align-items: center;
  border: 2px solid globals.$primary-background-color;
  cursor: pointer;
  display: flex;
  outline: none;
}

.primary {
  background-color: globals.$primary-background-color;

  &:hover,
  &:focus {
    background-color: globals.$primary-background-color-hover-and-focus;
    border: 2px solid
      globals.$primary-background-color-hover-and-focus;
  }

  &:active {
    background-color: globals.$primary-background-color-active;
    border: 2px solid globals.$primary-background-color-active;
  }
}

.secondary {
  background-color: white;
  border: 2px solid globals.$secondary-text-color;

  &:hover,
  &:focus {
    background-color: globals.$secondary-background-color-hover-and-active;
    border: 2px solid globals.$secondary-border-color-hover-and-focus;
    color: globals.$secondary-text-color-hover-and-focus;
  }

  &:active {
    background-color: globals.$secondary-background-color-active;
  }
}

.small {
  border-radius: 6px;
  height: 25.8px;
  padding: 4px 6px;

  .leadingIcon,
  .trailingIcon {
    height: 12px;
  }
}

.medium {
  border-radius: 8px;
  height: 34.5px;
  padding: 6px 9px;

  .leadingIcon,
  .trailingIcon {
    height: 16px;
  }
}

.large {
  border-radius: 10px;
  height: 43px;
  padding: 8px 12px;

  .leadingIcon,
  .trailingIcon {
    height: 20px;
  }
}

.smallRipple {
  border-radius: 6px;
}

.mediumRipple {
  border-radius: 8px;
}

.largeRipple {
  border-radius: 10px;
}

.typography {
  align-items: center;
  display: flex;
  gap: 6px;
}

.disableRipple {
  pointer-events: none !important;
}

.disabled {
  background-color: lightgray !important;
  border: 2px solid lightgray;
  cursor: default !important;
  user-select: none !important;
}
