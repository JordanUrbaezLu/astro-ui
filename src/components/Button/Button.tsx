import * as React from "react";
import styles from "./Button.module.scss";
import classNames from "classnames";
import { Typography } from "../Typography/Typography";
import { Ripple } from "../Ripple";

export type ButtonVariant = "primary" | "secondary";
export type ButtonSize = "small" | "medium" | "large";
export type ButtonType = "button" | "submit" | "reset";

export interface ButtonProps
  extends React.ComponentPropsWithoutRef<"button"> {
  /**
   * The content for the Button
   */
  children: string;
  /**
   * If the Button is disabled
   *
   * @default false
   */
  disabled?: boolean;
  /**
   * The leading icon for the Button
   */
  leadingIcon?: React.ReactNode;
  /**
   * The callback fired when the Button is clicked
   */
  onClick?: () => void;
  /**
   * The size of the Button
   *
   * @default "medium"
   */
  size?: ButtonSize;
  /**
   * The trailing icon for the Button
   */
  trailingIcon?: React.ReactNode;
  /**
   * The type of the Button
   */
  type?: ButtonType;
  /**
   * The variant of the Button
   *
   * @default "primary"
   */
  variant?: ButtonVariant;
}

/**
 * @public
 *
 * @description
 *
 * The Button component is used to allows user to take an action.
 */
export const Button = React.forwardRef<
  HTMLButtonElement,
  ButtonProps
>((props, ref) => {
  const {
    className,
    children,
    disabled = false,
    leadingIcon,
    onClick,
    size = "medium",
    trailingIcon,
    type,
    variant = "primary",
    ...rest
  } = props;

  return (
    <button
      className={classNames(
        className,
        styles.button,
        disabled && styles.disabled,
        size === "small"
          ? styles.small
          : size === "large"
          ? styles.large
          : styles.medium,
        variant === "secondary" ? styles.secondary : styles.primary
      )}
      disabled={disabled}
      onClick={onClick}
      ref={ref}
      type={
        type === "reset"
          ? "reset"
          : type === "submit"
          ? "submit"
          : "button"
      }
      {...rest}
    >
      {!disabled && <Ripple position="cursor" />}
      <Typography
        className={styles.typography}
        type={
          size === "large" ? "h1" : size === "medium" ? "h2" : "h3"
        }
        variant={variant}
      >
        {leadingIcon && (
          <span className={styles.leadingIcon}>{leadingIcon}</span>
        )}
        {children}
        {trailingIcon && (
          <span className={styles.trailingIcon}>{trailingIcon}</span>
        )}
      </Typography>
    </button>
  );
});

Button.displayName = "Button";
