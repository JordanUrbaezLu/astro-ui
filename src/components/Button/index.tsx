/* istanbul ignore file */
export * from "./Button";
