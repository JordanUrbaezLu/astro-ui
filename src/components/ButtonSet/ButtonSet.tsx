import * as React from "react";
import classNames from "classnames";
import styles from "./ButtonSet.module.scss";

export interface ButtonSetProps extends React.ComponentPropsWithoutRef<"div"> {
  /**
   * The buttons in the Button Set
   */
  children: React.ReactNode;
}

export const ButtonSet: React.FC<ButtonSetProps> = (props) => {
  const { className, children, ...rest } = props;

  return (
    <div className={classNames(styles.buttonSet, className)} {...rest}>
      {children}
    </div>
  );
};
