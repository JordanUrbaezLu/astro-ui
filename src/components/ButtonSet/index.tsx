/* istanbul ignore file */
export * from "./ButtonSet";
