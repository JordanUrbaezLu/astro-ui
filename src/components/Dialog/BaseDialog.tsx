import * as React from "react";
import styles from "./BaseDialog.module.scss";
import { IconButton } from "../IconButton/IconButton";
import { useOnClickOutside, useOnKeyDown } from "../../hooks";
import { TbX } from "react-icons/tb";
import { FocusTrap } from "../FocusTrap/FocusTrap";
import { Separator } from "../Separator/Separator";
import classNames from "classnames";
import { Typography } from "../Typography/Typography";

export type BaseDialogVariant =
  | "bottom"
  | "center"
  | "left"
  | "right";

export interface BaseDialogProps
  extends React.ComponentPropsWithoutRef<"div"> {
  /**
   * The accessible label for the Dialog
   */
  ariaLabel?: string;
  /**
   * The buttons for the Dialog
   */
  buttons?: React.ReactNode;
  /**
   * The content for the Dialog
   */
  content: string;
  /**
   * The callback fired when the Dialog closes
   */
  onClose: () => void;
  /**
   * The title for the Dialog
   */
  title: string;
  /**
   * The variant of the Dialog
   */
  variant?: BaseDialogVariant;
}

/**
 * @private
 */
export const BaseDialog: React.FC<BaseDialogProps> = (props) => {
  const {
    ariaLabel,
    buttons,
    className,
    content,
    onClose,
    title,
    variant,
    ...rest
  } = props;

  const dialogRef = React.useRef<HTMLDivElement>(null);

  useOnClickOutside(dialogRef, onClose);

  useOnKeyDown(["Escape"], onClose);

  return (
    <FocusTrap>
      <div
        aria-label={ariaLabel ?? "dialog"}
        className={classNames(
          className,
          styles.dialogContainer,
          variant === "bottom" && styles.bottom,
          variant === "center" && styles.center,
          variant === "left" && styles.left,
          variant === "right" && styles.right
        )}
        ref={dialogRef}
        role="dialog"
        {...rest}
      >
        <div className={styles.dialogHeader}>
          <Typography
            className={styles.dialogTitle}
            type="h1"
            variant="primary"
          >
            {title}
          </Typography>
          <IconButton
            className={styles.dialogIcon}
            content={<TbX />}
            onClick={onClose}
          />
        </div>
        {(variant === "left" || variant === "right") && (
          <Separator direction="horizontal" stroke={1} />
        )}
        <Typography
          className={styles.dialogContent}
          type="h2"
          variant="primary"
        >
          {content}
        </Typography>
        {buttons && (
          <div
            className={classNames(
              variant === "left" && styles.dialogButtonsContainer,
              variant === "right" && styles.dialogButtonsContainer
            )}
          >
            <Separator direction="horizontal" stroke={1} />
            <div className={styles.dialogButtons}>{buttons}</div>
          </div>
        )}
      </div>
    </FocusTrap>
  );
};

BaseDialog.displayName = "BaseDialog";
