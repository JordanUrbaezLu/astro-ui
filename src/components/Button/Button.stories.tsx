import { Story } from "@storybook/react";
import { Button, ButtonProps } from "./Button";
import {
  MdOutlineArrowBack,
  MdOutlineArrowForward,
} from "react-icons/md";
import { FlexRowContainer } from "../../../.storybook/components/FlexRowContainer";

export default {
  component: Button,
  title: "Components/Inputs/Button",
};

const Template: Story<ButtonProps> = (args) => {
  return <Button {...args}>Button</Button>;
};

export const Default = Template.bind({});

export const Size = () => {
  return (
    <FlexRowContainer>
      <Button size="small">Small</Button>
      <Button size="medium">Medium</Button>
      <Button size="large">Large</Button>
    </FlexRowContainer>
  );
};

export const Variant = () => {
  return (
    <FlexRowContainer>
      <Button variant="primary">Primary</Button>
      <Button variant="secondary">Secondary</Button>
    </FlexRowContainer>
  );
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

export const LeadingIcon = Template.bind({});
LeadingIcon.args = {
  leadingIcon: <MdOutlineArrowBack />,
};

export const TrailingIcon = Template.bind({});
TrailingIcon.args = {
  trailingIcon: <MdOutlineArrowForward />,
};

export const LeadingIconAndTrailingIcon = Template.bind({});
LeadingIconAndTrailingIcon.args = {
  leadingIcon: <MdOutlineArrowBack />,
  trailingIcon: <MdOutlineArrowForward />,
};

export const AllProps = Template.bind({});
AllProps.args = {
  size: "large",
  variant: "secondary",
  leadingIcon: <MdOutlineArrowBack />,
  trailingIcon: <MdOutlineArrowForward />,
};
