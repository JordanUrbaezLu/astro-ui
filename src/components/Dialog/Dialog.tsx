import * as React from "react";
import styles from "./Dialog.module.scss";
import { Backdrop } from "../Backdrop/Backdrop";
import { BaseDialog, BaseDialogProps } from "./BaseDialog";
import { useLockBodyScroll } from "../../hooks/useLockBodyScroll";
import { CSSTransition } from "react-transition-group";
import classNames from "classnames";

export interface DialogProps extends BaseDialogProps {
  /**
   * If the Dialog has a backdrop
   *
   * @default true
   */
  hasBackdrop?: boolean;
  /**
   * If the Dialog is open
   *
   * @default false
   */
  isOpen?: boolean;
  /**
   * The callback fired when the Dialog opens
   */
  onOpen: () => void;
  /**
   * The trigger for the Dialog
   */
  trigger: React.ReactElement;
}

/**
 * @public
 *
 * @description
 *
 * The Dialog component gives attention to a task a user can take.
 */
export const Dialog: React.FC<DialogProps> = (props) => {
  const {
    ariaLabel,
    buttons,
    className,
    content,
    hasBackdrop = true,
    isOpen = false,
    onClose,
    onOpen,
    title,
    trigger,
    variant = "center",
    ...rest
  } = props;

  useLockBodyScroll(isOpen);

  return (
    <>
      {React.cloneElement(trigger, {
        onClick: onOpen,
        role: "button",
        tabIndex: 0,
      })}
      <CSSTransition
        in={isOpen}
        timeout={250}
        mountOnEnter
        unmountOnExit
        classNames={{
          enter: styles.enter,
          enterActive: styles.enterActive,
          exit: styles.exit,
          exitActive: styles.exitActive,
        }}
      >
        <Backdrop showBackground={hasBackdrop}>
          <div
            className={classNames(
              styles.dialogBackground,
              variant === "bottom" && styles.bottom,
              variant === "center" && styles.center,
              variant === "left" && styles.left,
              variant === "right" && styles.right
            )}
          >
            <BaseDialog
              ariaLabel={ariaLabel}
              buttons={buttons}
              className={className}
              content={content}
              onClose={onClose}
              title={title}
              variant={variant}
              {...rest}
            />
          </div>
        </Backdrop>
      </CSSTransition>
    </>
  );
};

Dialog.displayName = "Dialog";
