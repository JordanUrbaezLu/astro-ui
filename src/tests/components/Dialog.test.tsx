import { fireEvent, render, screen } from "@testing-library/react";
import { Dialog } from "../../components/Dialog/Dialog";
import { Button } from "../../components/Button/Button";
import { ButtonSet } from "../../components/ButtonSet/ButtonSet";
import { axe } from "jest-axe";
import { BaseDialogVariant } from "src/components/Dialog/BaseDialog";

describe("Dialog", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <Dialog
        content="Dialog Content"
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
        title="Dialog Title"
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <Dialog
        content="Dialog Content"
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
        title="Dialog Title"
      />
    );
    const results = axe(container, {
      rules: { tabindex: { enabled: false } },
    });
    expect(await results).toHaveNoViolations();
  });
});

describe("Interaction", () => {
  test("Should trigger onClose on Escape key down.", () => {
    const onClose = jest.fn();

    render(
      <Dialog
        content="Dialog Content"
        isOpen
        onOpen={jest.fn}
        onClose={onClose}
        trigger={<Button>Trigger</Button>}
        title="Dialog Title"
      />
    );

    fireEvent.keyDown(document.body, { key: "Escape" });

    expect(onClose).toHaveBeenCalledTimes(1);
  });
});

describe("Props", () => {
  test("Should render buttons correctly.", () => {
    render(
      <Dialog
        buttons={
          <ButtonSet>
            <Button variant="secondary">Cancel</Button>
            <Button variant="secondary">Confirm</Button>
          </ButtonSet>
        }
        content="Dialog Content"
        isOpen
        onOpen={jest.fn()}
        onClose={jest.fn()}
        trigger={<Button variant="secondary">Open Dialog!</Button>}
        title="Dialog Title"
      />
    );
    expect(screen.getByText("Cancel")).toBeInTheDocument();
    expect(screen.getByText("Confirm")).toBeInTheDocument();
  });
  test("Should render content correctly.", () => {
    render(
      <Dialog
        content="Dialog Content"
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
        title="Dialog Title"
      />
    );
    expect(screen.getByText("Dialog Content")).toBeInTheDocument();
  });

  test("Should render closed correctly.", () => {
    render(
      <Dialog
        content="Dialog Content"
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
        title="Dialog Title"
      />
    );
    expect(
      screen.queryByText("Dialog Title")
    ).not.toBeInTheDocument();
  });

  test("Should render title correctly.", () => {
    render(
      <Dialog
        content="Dialog Content"
        isOpen
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
        title="Dialog Title"
      />
    );
    expect(screen.getByText("Dialog Title")).toBeInTheDocument();
  });

  test.each<BaseDialogVariant>(["bottom", "center", "left", "right"])(
    "Should render variants correctly.",
    (variant) => {
      render(
        <Dialog
          ariaLabel={variant}
          buttons={
            <ButtonSet>
              <Button>Cancel</Button>
              <Button>Confirm</Button>
            </ButtonSet>
          }
          content="Dialog Content"
          isOpen
          onOpen={jest.fn}
          onClose={jest.fn()}
          trigger={<Button>Trigger</Button>}
          title="Dialog Title"
          variant={variant}
        />
      );
      expect(screen.getByRole("dialog")).toHaveClass(variant);
    }
  );
});
