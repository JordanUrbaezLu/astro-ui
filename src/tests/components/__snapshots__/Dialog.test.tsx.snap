// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog Should render component correctly. 1`] = `
<button
  class="button medium primary"
  role="button"
  tabindex="0"
  type="button"
>
  <div
    class="container"
    role="presentation"
  />
  <div
    class="typography typography typographyh2 primary normal"
  >
    Trigger
  </div>
</button>
`;
